File: leetcode/2696-minimum-string-length-after-removing-substrings/solution.py
class Solution:
    def minLength(self, s: str) -> int:
        stack = []

        for c in s:
            if len(stack) == 0:
                stack.append(c)
                continue

            if c == 'B' and stack[-1] == 'A':
                stack.pop()
            elif c == 'D' and stack[-1] == 'C':
                stack.pop()
            else:
                stack.append(c)

        return len(stack)

File: leetcode/2696-minimum-string-length-after-removing-substrings/solution.cpp
/**
 * Author: 1chooo<hugo970217@gmail.com>
 * Problem: https://leetcode.com/problems/minimum-string-length-after-removing-substrings
 * Runtime: 3ms (89.08%)
 */

const static auto _ = []() {
    cin.tie(nullptr)->sync_with_stdio(false);
    return nullptr;
}();

class Solution {
public:
    int minLength(string s) {
        stack<char> stk;
        for (char c : s) {
            if (stk.empty()) {
                stk.push(c);
                continue;
            }

            if (c == 'B' && stk.top() == 'A') {
                stk.pop();
            } else if (c == 'D' && stk.top() == 'C') {
                stk.pop();
            } else {
                stk.push(c);
            }
        }
        return stk.size();
    }
};
